pkg/engine2/reconciler/reconciler.go

```diff
@@ -3,65 +3,195 @@ package reconciler
+			queue = appendToQueue(deleteRequest{resource: res, explicit: explicit}, queue)
+			// find deployment dependencies to ensure the resource wont get recreated
+			queue, err = addAllDeploymentDependencies(c, res, explicit, queue)
+			if err != nil {
+				errs = errors.Join(errs, err)
+			}
+
+		}
+		// try to cleanup, if the resource is removable
+		for res := range upstreams.Union(downstreams) {
+			queue = appendToQueue(deleteRequest{resource: res, explicit: false}, queue)
+		}
 	}
+	return errs
+}
 
-	err = construct.RemoveResource(op, resource)
-	if err != nil {
-		return err
+func appendToQueue(request deleteRequest, queue []deleteRequest) []deleteRequest {
+	for i, req := range queue {
+		if req.resource == request.resource && req.explicit == request.explicit {
+			return queue
+			// only update if it wasnt explicit previously and now it is
+		} else if req.resource == request.resource && !req.explicit {
+			queue[i] = request
+			return queue
+		} else if req.resource == request.resource && req.explicit {
+			return queue
+		}
 	}
+	return append(queue, request)
+}
 
-	namespacedResources, err := findAllResourcesInNamespace(c, resource)
+func addAllDeploymentDependencies(
+	ctx solution_context.SolutionContext,
+	resource construct.ResourceId,
+	explicit bool,
+	queue []deleteRequest,
+) ([]deleteRequest, error) {
+
+	deploymentDeps, err := knowledgebase.Upstream(
+		ctx.DeploymentGraph(),
+		ctx.KnowledgeBase(),
+		resource,
+		knowledgebase.ResourceDirectLayer,
+	)
 	if err != nil {
-		return err
-	}
-	for _, res := range namespacedResources.ToSlice() {
-		// Since we are explicitly deleting the namespace resource, we will explicitly delete the resources namespaced to it
-		errs = errors.Join(errs, RemoveResource(c, res, explicit))
+		return nil, err
 	}
-	// try to cleanup, if the resource is removable
-	for res := range upstreams.Union(downstreams) {
-		errs = errors.Join(errs, RemoveResource(c, res, false))
+
+	for _, dep := range deploymentDeps {
+		res, err := ctx.RawView().Vertex(dep)
+		if err != nil {
+			return nil, err
+		}
+		rt, err := ctx.KnowledgeBase().GetResourceTemplate(dep)
+		if err != nil {
+			return nil, err
+		}
+		if collectionutil.Contains(queue, deleteRequest{resource: dep, explicit: explicit}) {
+			continue
+		}
+
+		shouldDelete := false
+
+		// check if the dep exists as a property on the resource
+		err = rt.LoopProperties(res, func(p knowledgebase.Property) error {
+			propVal, err := res.GetProperty(p.Details().Path)
+			if err != nil {
+				return err
+			}
+			found := false
+			switch val := propVal.(type) {
+			case construct.ResourceId:
+				if val == resource {
+					found = true
+				}
+			case construct.PropertyRef:
+				if val.Resource == resource {
+					found = true
+				}
+			default:
+				if reflect.ValueOf(val).Kind() == reflect.Slice || reflect.ValueOf(val).Kind() == reflect.Array {
```

Comment:
This doesn't look deeper into maps, or lists of lists but that can be a `TODO`. Only case it might be relevant is for policies that have a PropertyRef.

Reply:
well that should be ok because doesnt loop properties get all indices of lists and keys of maps?

Reply:
You're right, but then shouldn't this not need to be added since it'd go into each index?